Fix panic=abort tests on fuchsia

c6c7 opened the pull request:
This PR goes one step further than #127594 by adding an unstable feature to make determining whether a Fuchsia process aborted more robust. There is certainly more work that needs to go in to landing the unstable feature, but I'm drafting this PR as a means to convey intent for the feature to survey others' thoughts.

r​? @tmandry

c6c7 commented:
r? @tmandry

c6c7 commented:
I should state explicitly that a goal of this PR is to enable Fuchsia testing for `tests/ui/process/signal-exit-status.rs `without needing to copy-paste the magic Fuchsia status code that also appears in `library/test/src/test_result.rs` (and any futures authors needing to determine whether a Fuchsia process was aborted).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9885972762/job/27304917708) [(plain)](https://github.com/rust-lang/rust/commit/130612bd5eabda5f04c9c031da48ef40fbf7b3ce/checks/27304917708/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.413   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.426 Collecting boolean-py==4.0
#13 0.429   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.441 Collecting chardet==5.1.0
---
#13 4.437 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.947 Collecting virtualenv
#13 4.983   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.054      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 81.9 MB/s eta 0:00:00
#13 5.105 Collecting platformdirs<5,>=3.9.1
#13 5.108   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.125 Collecting distlib<1,>=0.3.7
#13 5.129   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.166 Collecting filelock<4,>=3.12.2
#13 5.170   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.170   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.259 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.431 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      202688 kB
DirectMap2M:     7137280 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/e1f45a1442bb373429b6b0838e8ef806ddddfeef/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-e1f45a1442bb373429b6b0838e8ef806ddddfeef-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5248 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_feature/src/unstable.rs:479: no tracking issue for feature fuchsia_exit_status_aborted
##[error]tidy error: /checkout/compiler/rustc_feature/src/unstable.rs:481: feature freeze_impls is not sorted by feature name (should be between fn_delegation and fuchsia_exit_status_aborted)
Expected a gate test for the feature 'fuchsia_exit_status_aborted'.
Hint: create a failing test file named 'tests/ui/feature-gates/feature-gate-fuchsia-exit-status-aborted.rs',
      with its failures due to missing usage of `#![feature(fuchsia_exit_status_aborted)]`.
Hint: If you already have such a test and don't want to rename it,
      you can also add a // gate-test-fuchsia_exit_status_aborted line to the test file.
tidy error: Found 1 features without a gate test.
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.1)
  Downloading pip-24.1.2-py3-none-any.whl.metadata (3.6 kB)
Downloading pip-24.1.2-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 77.9 MB/s eta 0:00:00
Installing collected packages: pip
---
checking C++ file formatting
some tidy checks failed


Command BootstrapCommand { command: LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustc" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "4" "--extra-checks=py:lint,cpp:fmt", failure_behavior: DelayFail, stdout: Print, stderr: Print, run_always: false } did not execute successfully.

Build completed unsuccessfully in 0:01:15
  local time: Thu Jul 11 05:30:33 UTC 2024
  network time: Thu, 11 Jul 2024 05:30:34 GMT
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9886054945/job/27305132626) [(plain)](https://github.com/rust-lang/rust/commit/ba47433c0294d50a08690dc452926328dba3f0a1/checks/27305132626/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.423   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.441 Collecting boolean-py==4.0
#13 0.448   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.463 Collecting chardet==5.1.0
---
#13 3.555 Building wheels for collected packages: reuse
#13 3.556   Building wheel for reuse (pyproject.toml): started
#13 3.880   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.881   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.882   Stored in directory: /tmp/pip-ephem-wheel-cache-u8kvvt6x/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.884 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.905   Attempting uninstall: setuptools
#13 3.906     Found existing installation: setuptools 59.6.0
#13 3.907     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.907     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.908     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.578 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.579 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.092 Collecting virtualenv
#13 5.140   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.299      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 36.3 MB/s eta 0:00:00
#13 5.351 Collecting platformdirs<5,>=3.9.1
#13 5.358   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.380 Collecting distlib<1,>=0.3.7
#13 5.388   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.435 Collecting filelock<4,>=3.12.2
#13 5.441   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.441   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.529 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.714 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 5.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      194496 kB
DirectMap2M:     8194048 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/8c39ac9eccdfaeff11bd3b6107817a81fe0a3a43/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-8c39ac9eccdfaeff11bd3b6107817a81fe0a3a43-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
error: expected identifier, found keyword `as`
##[error]   --> /checkout/library/test/src/test_result.rs:103:28
    |
103 |         Some(0xC0000409u32 as i32) => TestResult::TrFailed,


error: expected one of `)`, `,`, `...`, `..=`, `..`, or `|`, found keyword `as`
    |
    |
103 |         Some(0xC0000409u32 as i32) => TestResult::TrFailed,
    |                           -^^ expected one of `)`, `,`, `...`, `..=`, `..`, or `|`
    |                           help: missing `,`


error: expected one of `)`, `,`, or `|`, found `i32`
    |
    |
103 |         Some(0xC0000409u32 as i32) => TestResult::TrFailed,
    |                              -^^^ expected one of `)`, `,`, or `|`
    |                              help: missing `,`


fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/future/async_drop.rs" "/checkout/library/core/src/future/pending.rs" "/checkout/library/core/src/alloc/mod.rs" "/checkout/library/core/src/alloc/global.rs" "/checkout/library/core/src/alloc/layout.rs" "/checkout/library/core/src/borrow.rs" "/checkout/library/core/src/pin.rs" "/checkout/library/core/src/lib.rs" "/checkout/library/core/src/option.rs" "/checkout/library/core/src/clone.rs" "/checkout/library/core/src/panic/location.rs" "/checkout/library/core/src/panic/panic_info.rs" "/checkout/library/core/src/panic/unwind_safe.rs" "/checkout/library/core/src/primitive_docs.rs" "/checkout/library/core/src/str/pattern.rs" "/checkout/library/core/src/cell/once.rs" "/checkout/library/core/src/str/error.rs" "/checkout/library/core/src/cell/lazy.rs" "/checkout/library/core/src/str/mod.rs" "/checkout/library/core/src/str/traits.rs" "/checkout/library/core/src/str/count.rs" "/checkout/library/core/src/str/validations.rs" "/checkout/library/core/src/str/lossy.rs" "/checkout/library/core/src/str/iter.rs" "/checkout/library/core/src/str/converts.rs" "/checkout/library/core/src/escape.rs" "/checkout/library/core/src/arch.rs" "/checkout/library/rustc-std-workspace-alloc/lib.rs" "/checkout/library/core/src/task/ready.rs" "/checkout/library/core/src/task/wake.rs" "/checkout/library/core/src/task/mod.rs" "/checkout/library/core/src/task/poll.rs" "/checkout/library/core/src/macros/mod.rs" "/checkout/library/test/src/cli.rs" "/checkout/library/test/src/term.rs" "/checkout/library/test/src/tests.rs" "/checkout/library/test/src/test_result.rs" "/checkout/library/test/src/formatters/json.rs" "/checkout/library/test/src/formatters/mod.rs" "/checkout/library/test/src/formatters/junit.rs" "/checkout/library/test/src/formatters/pretty.rs" "/checkout/library/test/src/formatters/terse.rs" "/checkout/library/core/src/iter/sources/from_fn.rs" "/checkout/library/core/src/iter/sources/from_coroutine.rs" "/checkout/library/core/src/iter/sources/repeat_n.rs" "/checkout/library/core/src/iter/sources/repeat_with.rs" "/checkout/library/core/src/iter/sources/once_with.rs" "/checkout/library/core/src/iter/sources/successors.rs" "/checkout/library/core/src/iter/sources/repeat.rs" "/checkout/library/core/src/iter/sources/once.rs" "/checkout/library/core/src/iter/sources/empty.rs" "/checkout/library/core/src/iter/mod.rs" "/checkout/library/test/src/helpers/metrics.rs" "/checkout/library/test/src/helpers/mod.rs" "/checkout/library/test/src/helpers/shuffle.rs" "/checkout/library/test/src/helpers/concurrency.rs" "/checkout/library/test/src/bench.rs" "/checkout/library/test/src/lib.rs" "/checkout/library/test/src/stats/tests.rs" "/checkout/library/test/src/console.rs" "/checkout/library/test/src/stats.rs" "/checkout/library/test/src/time.rs" "/checkout/library/test/src/options.rs" "/checkout/library/core/src/future/into_future.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Jul 11 05:37:41 UTC 2024
  network time: Thu, 11 Jul 2024 05:37:41 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

c6c7 commented:
#58590 references #102032 which ignored some tests because of the inability of those tests easily recognize Fuchsia's process status code in various error cases. This change should consider re-enabling testing for Fuchsia in those cases.

Nilstrieb commented on compiler/rustc_feature/src/unstable.rs:
```diff
@@ -475,6 +475,8 @@ declare_features! (
     /// Support delegating implementation of functions to other already implemented functions.
     (incomplete, fn_delegation, "1.76.0", Some(118212)),
+    /// Support ExitStatusExt::aborted_code on Fuchsia platform.
+    (unstable, fuchsia_exit_status_aborted, "1.79.0", None),
```
you don't need to touch the compiler, you can just add library features

c6c7 replied:
Ah, ok. I was led astray by https://rustc-dev-guide.rust-lang.org/implementing_new_features.html. That makes sense though. I see now that this not a language feature.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9886342908/job/27305955477) [(plain)](https://github.com/rust-lang/rust/commit/935346330aa538a7edcb4be7c1d0e7292afa6283/checks/27305955477/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.402   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.414 Collecting boolean-py==4.0
#13 0.417   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.429 Collecting chardet==5.1.0
---
#13 4.936   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.034      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 59.1 MB/s eta 0:00:00
#13 5.088 Collecting filelock<4,>=3.12.2
#13 5.092   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.108 Collecting distlib<1,>=0.3.7
#13 5.111   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.117      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 118.8 MB/s eta 0:00:00
#13 5.145 Collecting platformdirs<5,>=3.9.1
#13 5.148   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.237 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.413 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 5.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      194496 kB
DirectMap2M:     6096896 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/8c39ac9eccdfaeff11bd3b6107817a81fe0a3a43/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-8c39ac9eccdfaeff11bd3b6107817a81fe0a3a43-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 28.54s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/test/src/test_result.rs at line 100:
         // On Windows we use __fastfail to abort, which is documented to use this
         // exception code.
         #[cfg(windows)]
-        Some(const {0xC0000409u32 as i32}) => TestResult::TrFailed,
+        Some(const { 0xC0000409u32 as i32 }) => TestResult::TrFailed,
         #[cfg(any(windows, unix))]
         Some(code) => TestResult::TrFailedMsg(format!("got unexpected return code {code}")),
         #[cfg(not(any(windows, unix)))]
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/sync/exclusive.rs" "/checkout/library/core/src/sync/mod.rs" "/checkout/library/core/src/sync/atomic.rs" "/checkout/library/rustc-std-workspace-alloc/lib.rs" "/checkout/library/core/src/fmt/num.rs" "/checkout/library/core/src/fmt/rt.rs" "/checkout/library/core/src/fmt/mod.rs" "/checkout/library/core/src/fmt/nofloat.rs" "/checkout/library/core/src/fmt/float.rs" "/checkout/library/core/src/fmt/builders.rs" "/checkout/library/core/src/mem/manually_drop.rs" "/checkout/library/core/src/mem/transmutability.rs" "/checkout/library/core/src/mem/mod.rs" "/checkout/library/core/src/mem/maybe_uninit.rs" "/checkout/library/core/src/intrinsics.rs" "/checkout/library/core/src/ub_checks.rs" "/checkout/library/test/src/cli.rs" "/checkout/library/test/src/term.rs" "/checkout/library/test/src/tests.rs" "/checkout/library/test/src/test_result.rs" "/checkout/library/core/src/array/ascii.rs" "/checkout/library/core/src/array/drain.rs" "/checkout/library/core/src/array/mod.rs" "/checkout/library/core/src/array/equality.rs" "/checkout/library/core/src/array/iter.rs" "/checkout/library/test/src/formatters/json.rs" "/checkout/library/test/src/formatters/mod.rs" "/checkout/library/test/src/formatters/junit.rs" "/checkout/library/test/src/formatters/pretty.rs" "/checkout/library/test/src/formatters/terse.rs" "/checkout/library/core/src/intrinsics/simd.rs" "/checkout/library/core/src/intrinsics/mir.rs" "/checkout/library/core/src/hash/sip.rs" "/checkout/library/core/src/hash/mod.rs" "/checkout/library/test/src/helpers/metrics.rs" "/checkout/library/test/src/helpers/mod.rs" "/checkout/library/test/src/helpers/shuffle.rs" "/checkout/library/test/src/helpers/concurrency.rs" "/checkout/library/test/src/bench.rs" "/checkout/library/test/src/lib.rs" "/checkout/library/test/src/stats/tests.rs" "/checkout/library/test/src/console.rs" "/checkout/library/test/src/stats.rs" "/checkout/library/test/src/time.rs" "/checkout/library/test/src/options.rs" "/checkout/library/test/src/event.rs" "/checkout/library/test/src/types.rs" "/checkout/library/core/src/alloc/mod.rs" "/checkout/library/core/src/alloc/global.rs" "/checkout/library/core/src/alloc/layout.rs" "/checkout/library/core/src/borrow.rs" "/checkout/library/core/src/pin.rs" "/checkout/library/core/src/lib.rs" "/checkout/library/core/src/option.rs" "/checkout/library/core/src/clone.rs" "/checkout/library/test/src/term/win.rs" "/checkout/library/test/src/term/terminfo/mod.rs" "/checkout/library/core/src/str/pattern.rs" "/checkout/library/core/src/str/error.rs" "/checkout/library/core/src/str/mod.rs" "/checkout/library/core/src/str/traits.rs" "/checkout/library/core/src/str/count.rs" "/checkout/library/core/src/str/validations.rs" "/checkout/library/core/src/arch.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Jul 11 06:09:58 UTC 2024
  network time: Thu, 11 Jul 2024 06:09:58 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

tgross35 commented on library/std/src/os/fuchsia/process.rs:
```diff
@@ -0,0 +1,25 @@
+use crate::process;
+use crate::sealed::Sealed;
+
+#[unstable(feature = "fuchsia_exit_status_aborted", issue = "none")]
```
Maybe just `fuchsia_exit_status`? In case there is eventually more than just `aborted_code`

c6c7 replied:
Good point. I changed the feature name to `fuchsia_exit_status`.

tgross35 commented on library/std/src/os/fuchsia/mod.rs:
```diff
@@ -3,4 +3,6 @@
 #![stable(feature = "raw_ext", since = "1.1.0")]
 
 pub mod fs;
+#[unstable(feature = "fuchsia_exit_status_aborted", issue = "none")]
+pub mod process;
```
Looks like you changed to `fuchsia_exit_status` in `process.rs`, but not here or the test

c6c7 replied:
You're quick! I just caught that as well and already updated it.

c6c7 committed:
Add ExitStatusExt::task_retcode for Fuchsia platform

The method task_retcode() method improves the ergonomics of
determining if a Fuchsia process was killed by the kernel and
for what reason.

c6c7 commented:
After a lot of thought, I've changed the approach of this PR significantly. In short, Zircon does not implement signals as the definition of the C Standard Library expects for the implementation of libc::abort(). As a result, I think it's impossible to expose an API on top of a return code from Zircon to determine if a task was aborted or not. The best we can do is smooth the path for retrieving a ZX_TASK_RETCODE.

I will be pushing a new version of this PR shortly.

c6c7 commented:
The longer story with citations....

There are cases in this repository that attempt to detect whether a process aborted or not. The first is in `fn get_result_from_exit_code(..., status: ExitStatus, ...) -> TestResult` (defined in `test::test_result`) which returns whether a test failed based on an `ExitStatus`. If a platform implements `libc::abort` abort as defined in ISO C, a panic in a test will result in calling `libc::abort` which will call `libc::raise(SIGABRT)`. Thus, `get_result_from_exit_code` will return a test failed when `ExitStatus::signal` returns `Some(SIGABRT)`. The story does not end here though because Fuchsia and Windows don't support signals, and thus `ExitStatus::signal` always returns `None` on those platforms.

(For Windows, it's apparently sufficient to check for `STATUS_FAIL_FAST_EXCEPTION`, but I don't know the details beyond the existing comments in `test::test_result`)

For Fuchsia, `libc::abort` does not call `libc::raise(SIGABRT)` because Fuchsia in this case deviates from the ISO C definition of `libc::abort`. Instead, Fuchsia executes the LLVM intrinsic `__builtin_trap()`[1] which generally raises an Zircon kernel exception. Unless the process installed an exception handler, the exception will go uncaught, and the kernel will kill the process with the code `ZX_TASK_RETCODE_EXCEPTION_KILL`. Thus, for `get_result_from_exit_code`, the best we can do to check if a process aborted is to observe the return code `ZX_TASK_RETCODE_EXCEPTION_KILL`.

The second case is `tests/ui/process/signal-exit-status.rs` which tests whether a process aborted as expected upon calling `core::intrinsics::abort()`. The goal is very similar to `get_result_from_exit_code`, but the result is not a `TestResult`. For Fuchsia, the hindrance for writing this test is not knowing what the return code from the process should be. My first thought was, "Oh, we should just implement `ExitStatus::aborted_code(&self) -> Option<i32>`." But this is fraught because of *there is no return code in Fuchsia that is set if and only if a process aborted.* The closest approximation to detecting if a process aborted is to check if the process returned `ZX_TASK_RETCODE_EXCEPTION_KILL`.

Taking a step back though, it's not actually the primary goal of `signal-exit-status` to determine that the process returned `ZX_TASK_RETCODE_EXCEPTION_KILL` and *no other* possible uncaught exception was raised. There is some potential for a test flake where the Zircon kernel during the `signal-exit-status` test raises an exception for a reason other than the `abort()` call; this seems very unlikely though. Thus, it's sufficient for `signal-exit-status` to merely check for `ZX_TASK_RETCODE_EXCEPTION_KILL`.

[1]: https://cs.opensource.google/fuchsia/fuchsia/+/main:zircon/system/ulib/c/stdlib/abort.cc;l=13;drc=acd4b64f7b555e9d9ef4d41ab5bd0e8927132853

c6c7 commented:
The latest change modifies the approach to implement `task_retcode()` instead of `aborted_code()`. I believe this API more directly corresponds to the semantics around Fuchsia processes and correctly requires callers to reason about whether the ZX_TASK_RETCODE indicates the condition they expect to check.

c6c7 commented on library/std/src/os/fuchsia/process.rs:
```diff
@@ -0,0 +1,52 @@
+use crate::process;
+use crate::sealed::Sealed;
+
+#[unstable(feature = "fuchsia_exit_status", issue = "none")]
+pub type zx_status_t = i32;
+#[unstable(feature = "fuchsia_exit_status", issue = "none")]
+pub const ZX_TASK_RETCODE_SYSCALL_KILL: zx_status_t = -1024;
+#[unstable(feature = "fuchsia_exit_status", issue = "none")]
+pub const ZX_TASK_RETCODE_OOM_KILL: zx_status_t = -1025;
+#[unstable(feature = "fuchsia_exit_status", issue = "none")]
+pub const ZX_TASK_RETCODE_POLICY_KILL: zx_status_t = -1026;
+#[unstable(feature = "fuchsia_exit_status", issue = "none")]
+pub const ZX_TASK_RETCODE_VDSO_KILL: zx_status_t = -1027;
+/// On Zircon (the Fuchsia kernel), an abort from userspace calls the
+/// LLVM implementation of __builtin_trap(), e.g., ud2 on x86, which
+/// raises a kernel exception. If a userspace process does not
+/// otherwise arrange exception handling, the kernel kills the process
+/// with this return code.
+#[unstable(feature = "fuchsia_exit_status", issue = "none")]
+pub const ZX_TASK_RETCODE_EXCEPTION_KILL: zx_status_t = -1028;
+#[unstable(feature = "fuchsia_exit_status", issue = "none")]
+pub const ZX_TASK_RETCODE_CRITICAL_PROCESS_KILL: zx_status_t = -1029;
```
I'm not thrilled about adding these constants here since `std::sys::pal::unix::process::zircon` exists. However,   since `std::sys::pal::unix::process::zircon` is private, I didn't want to cross the bridge of exposing parts of that module in this change.

tgross35 replied:
Could/should these be added to `libc` instead?

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9945791167/job/27474876718) [(plain)](https://github.com/rust-lang/rust/commit/f98bc4b3b6441a3f47da95abad754b8b5f7e13a8/checks/27474876718/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:0b2256b8c012f0828dc542b3febcab082c67f72b)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.451   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.468 Collecting boolean-py==4.0
#13 0.474   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.490 Collecting chardet==5.1.0
---
#13 3.762 Building wheels for collected packages: reuse
#13 3.764   Building wheel for reuse (pyproject.toml): started
#13 4.103   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 4.104   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 4.104   Stored in directory: /tmp/pip-ephem-wheel-cache-sy0xqjqd/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 4.107 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 4.131   Attempting uninstall: setuptools
#13 4.131     Found existing installation: setuptools 59.6.0
#13 4.132     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.132     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 4.133     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.829 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.829 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.379 Collecting virtualenv
#13 5.429   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 5.639      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 27.4 MB/s eta 0:00:00
#13 5.682 Collecting distlib<1,>=0.3.7
#13 5.688   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.735 Collecting filelock<4,>=3.12.2
#13 5.742   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.742   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 5.773 Collecting platformdirs<5,>=3.9.1
#13 5.780   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.869 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 6.070 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 6.2s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      196544 kB
DirectMap2M:     7143424 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/eb72697e41b00e5d8723f14c64a969d59d9b9474/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-eb72697e41b00e5d8723f14c64a969d59d9b9474-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 28.88s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/std/src/sys/pal/unix/process/zircon.rs at line 36:
     T: TryInto<zx_status_t> + Copy,
 {
     if let Ok(status) = TryInto::try_into(t) {
-        if status < 0 {
-            Err(io::Error::from_raw_os_error(status))
-            Ok(t)
-        }
-        }
+        if status < 0 { Err(io::Error::from_raw_os_error(status)) } else { Ok(t) }
         Err(io::Error::last_os_error())
     }
     }
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/tests/run-make/output-filename-conflicts-with-directory/foo.rs" "/checkout/library/std/src/sys/pal/wasi/env.rs" "/checkout/tests/run-make/proc-macro-init-order/a.rs" "/checkout/tests/run-make/proc-macro-init-order/rmake.rs" "/checkout/tests/run-make/proc-macro-init-order/b.rs" "/checkout/tests/run-make/proc-macro-init-order/c.rs" "/checkout/library/std/src/sys/pal/teeos/os.rs" "/checkout/library/std/src/sys/pal/teeos/stdio.rs" "/checkout/library/std/src/sys/pal/teeos/net.rs" "/checkout/library/std/src/sys/pal/teeos/rand.rs" "/checkout/library/std/src/sys/pal/teeos/alloc.rs" "/checkout/library/std/src/sys/pal/teeos/thread.rs" "/checkout/library/std/src/sys/pal/teeos/mod.rs" "/checkout/library/std/src/sys/pal/wasip2/cabi_realloc.rs" "/checkout/library/std/src/sys/pal/wasip2/mod.rs" "/checkout/tests/run-make/manual-link/main.rs" "/checkout/tests/run-make/manual-link/foo.rs" "/checkout/library/std/src/sys/pal/uefi/helpers.rs" "/checkout/library/std/src/sys/pal/uefi/os.rs" "/checkout/library/std/src/sys/pal/uefi/stdio.rs" "/checkout/tests/run-make/interdependent-c-libraries/bar.rs" "/checkout/library/std/src/sys/pal/uefi/time.rs" "/checkout/tests/run-make/interdependent-c-libraries/main.rs" "/checkout/library/std/src/sys/pal/uefi/alloc.rs" "/checkout/tests/run-make/interdependent-c-libraries/foo.rs" "/checkout/library/std/src/sys/pal/uefi/thread.rs" "/checkout/library/std/src/sys/pal/uefi/tests.rs" "/checkout/library/std/src/sys/pal/uefi/args.rs" "/checkout/library/std/src/sys/pal/uefi/mod.rs" "/checkout/library/std/src/sys/pal/uefi/env.rs" "/checkout/tests/run-make/static-pie/rmake.rs" "/checkout/tests/run-make/static-pie/test-aslr.rs" "/checkout/library/std/src/sys/pal/wasm/atomics/futex.rs" "/checkout/library/std/src/sys/pal/wasm/atomics/thread.rs" "/checkout/tests/run-make/pretty-print-with-dep-file/rmake.rs" "/checkout/library/std/src/sys/pal/wasm/alloc.rs" "/checkout/library/std/src/sys/pal/wasm/mod.rs" "/checkout/tests/run-make/pretty-print-with-dep-file/with-dep.rs" "/checkout/library/std/src/sys/pal/wasm/env.rs" "/checkout/tests/run-make/c-link-to-rust-dylib/rmake.rs" "/checkout/tests/run-make/c-link-to-rust-dylib/foo.rs" "/checkout/tests/run-make/cross-lang-lto-riscv-abi/rmake.rs" "/checkout/tests/run-make/cross-lang-lto-riscv-abi/riscv-xlto.rs" "/checkout/tests/run-make/wasm-symbols-not-exported/bar.rs" "/checkout/tests/run-make/wasm-symbols-not-exported/rmake.rs" "/checkout/tests/run-make/wasm-symbols-not-exported/foo.rs" "/checkout/tests/run-make/panic-impl-transitive/rmake.rs" "/checkout/tests/run-make/panic-impl-transitive/panic-impl-provider.rs" "/checkout/tests/run-make/panic-impl-transitive/panic-impl-consumer.rs" "/checkout/library/std/src/sys/pal/unix/l4re.rs" "/checkout/library/std/src/sys/pal/unix/os.rs" "/checkout/tests/run-make/ice-dep-cannot-find-dep/a.rs" "/checkout/tests/run-make/ice-dep-cannot-find-dep/rmake.rs" "/checkout/tests/run-make/ice-dep-cannot-find-dep/b.rs" "/checkout/tests/run-make/ice-dep-cannot-find-dep/c.rs" "/checkout/library/std/src/sys/pal/unix/process/process_fuchsia.rs" "/checkout/tests/run-make/libtest-padding/rmake.rs" "/checkout/tests/run-make/libtest-padding/tests.rs" "/checkout/library/std/src/sys/pal/unix/process/process_unix/tests.rs" "/checkout/library/std/src/sys/pal/unix/process/process_unsupported/wait_status/tests.rs" "/checkout/library/std/src/sys/pal/unix/process/process_unsupported/wait_status.rs" "/checkout/library/std/src/sys/pal/unix/process/process_unix.rs" "/checkout/library/std/src/sys/pal/unix/process/zircon.rs" "/checkout/tests/run-make/output-filename-conflicts-with-directory/rmake.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Mon Jul 15 20:03:34 UTC 2024
  network time: Mon, 15 Jul 2024 20:03:34 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #125443) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tmandry commented:
LGTM but I'm not on Fuchsia anymore, so unassigning myself. Someone on the library team should review this – @c6c7, use `r? library` once this is no longer a draft.
